merlion/models/utils/rolling_window_dataset.py

```diff
@@ -47,168 +60,70 @@ def __init__(
             self.data = data_df.values
             self.timestamp = data_df.index
-            self.target = data_df.iloc[:, target_seq_index].values if self._label_axis == 0 else None
-            self.target_timestamp = data_df.index
+            self.target = self.data if self.autoregressive else self.data[:, target_seq_index]
 
-        self.shuffle = shuffle
-        self._data_len = len(data)
+    @property
+    def autoregressive(self):
+        return self.target_seq_index is None
 
     def __len__(self):
-        return self._data_len
+        if self.autoregressive:
+            return len(self.data) - self.n_past
+        else:
+            return len(self.data) - self.n_past - self.n_future
 
     def __iter__(self):
-
-        if self.batch_size is None or self.batch_size < 1:
-            if self._label_axis == 0:
-                yield self._get_entire_train_data_along_time()
-            elif self._label_axis == 1:
-                yield self._get_entire_train_data_along_sequence()
-            return
-
-        if self.batch_size == 1:
-            yield from self._get_iterator()
-        elif self.batch_size > 1:
-            batch = list()
-            for i in self._get_iterator():
-                batch.append(i)
-                if len(batch) >= self.batch_size:
-                    yield batch
-                    batch = list()
-
-    def _get_iterator(self):
-        """
-        For example, if the input is like the following
-
-            for self._label_axis = 0, we roll the window along the time axis
-            .. code-block:: python
-
-                data = (t0, (1,2,3)),
-                       (t1, (4,5,6)),
-                       (t2, (7,8,9)),
-                       (t3, (10,11,12))
-                target_seq_index = 0
-                n_past = 2
-                n_future = 2
-
-        Then we may yield the following for the first time
-
-            .. code-block:: python
-                (
-                    (t0, (1,2,3)),
-                    (t1, (4,5,6))
-                ),
-                (
-                    (t2, (7))
-                    (t3, (10))
-                )
-
-        """
-        if self._label_axis == 0:
-            _valid_rolling_steps = len(self) - self.n_past - self.n_future
-        elif self._label_axis == 1:
-            _valid_rolling_steps = len(self) - self.n_past
-        order = np.random.permutation(_valid_rolling_steps + 1) if self.shuffle \
-            else range(_valid_rolling_steps + 1)
+        batch = []
+        order = np.random.permutation(len(self)) if self.shuffle else range(len(self))
         for i in order:
-            yield self[i]
+            batch.append(self[i])
+            if self.batch_size is not None and len(batch) >= self.batch_size:
+                yield self.collate_batch(batch)
+                batch = []
+        if len(batch) > 0:
+            yield self.collate_batch(batch)
```

Comment:
the only concern for me is the efficiency for very large batch size, for example, for batch_size=None. The previous implementation, although look redundant, I wanted to give a pre-assigned numpy array instead of using python list append in this case.

Reply:
If the collate function were incrementally doing `np.concatenate()`, I would agree with your concern. However in this case, we are only calling `np.stack()` once, which is quite efficient.

Reply:
I was not worried about the stack() as it is one time operation. I am not pretty sure about the performance of adding new elements using python list append() vs a pre-allocated numpy array indexing update, there are a few discussions but I am not quite sure.

Reply:
append() is appending reference object, shall be fine.